cmd/crypto/vault.go

```diff
@@ -130,70 +129,80 @@ func NewVault(config VaultConfig) (KMS, error) {
-// intervals until successful.
-func (v *vaultService) reauthenticate() {
-	retryDelay := 1 * time.Minute
-	go func() {
-		for {
-			if err := v.authenticate(); err != nil {
-				time.Sleep(retryDelay)
-				continue
-			}
-			return
-		}
-	}()
-}
-
-// renewer calls vault client's renewer that automatically
-// renews secret periodically
-func (v *vaultService) renewer(secret *vault.Secret) {
+// renewSecret tries to renew the given secret. It blocks
+// until it receives either the new secret or encounters an error.
+func (v *vaultService) renewSecret(secret *vault.Secret) (*vault.Secret, error) {
 	renewer, err := v.client.NewRenewer(&vault.RenewerInput{
 		Secret: secret,
 	})
 	if err != nil {
 		logger.FatalIf(err, "crypto: hashicorp vault token renewer could not be started")
 	}
-	v.tokenRenewer = renewer
 	go renewer.Renew()
 	defer renewer.Stop()
 
 	for {
 		select {
 		case err := <-renewer.DoneCh():
 			if err != nil {
-				v.reauthenticate()
-				renewer.Stop()
-				return
+				return nil, err
 			}
-
-			// Renewal is now over
-		case renewal := <-renewer.RenewCh():
-			v.leaseDuration = time.Duration(renewal.Secret.Auth.LeaseDuration)
+		case renew := <-renewer.RenewCh():
+			if renew.Secret == nil || renew.Secret.Auth == nil {
+				return nil, ErrKMSAuthLogin
+			}
+			return renew.Secret, nil
 		}
 	}
 }
 
-// authenticate logs the app to vault, and starts the auto renewer
-// before secret expires
-func (v *vaultService) authenticate() (err error) {
+// login tries to authenticate the minio server to
+// the Vault KMS using the approle ID and secret.
+func (v *vaultService) login() (*vault.Secret, error) {
 	payload := map[string]interface{}{
 		"role_id":   v.config.Auth.AppRole.ID,
 		"secret_id": v.config.Auth.AppRole.Secret,
 	}
-	var secret *vault.Secret
-	secret, err = v.client.Logical().Write("auth/approle/login", payload)
+	secret, err := v.client.Logical().Write("auth/approle/login", payload)
 	if err != nil {
-		return
+		return nil, err
+	}
+	if secret == nil || secret.Auth == nil {
+		return nil, ErrKMSAuthLogin
 	}
-	if secret.Auth == nil {
-		err = ErrKMSAuthLogin
-		return
+	return secret, nil
+}
+
+// authenticate tries to authenticate the minio server
+// to the Vault KMS and starts a background job to renew
+// the login.
+func (v *vaultService) authenticate() error {
+	secret, err := v.login()
+	if err != nil {
+		return err
 	}
 	v.client.SetToken(secret.Auth.ClientToken)
 	v.leaseDuration = time.Duration(secret.Auth.LeaseDuration)
-	go v.renewer(secret)
-	return
+
+	// Start background job trying to renew the token
+	// or (if this fails) try to login again with app-ID and app-Secret.
+	go func(secret *vault.Secret) {
+		var err error
+		for {
+			secret, err = v.renewSecret(secret) // try to renew the secret (blocking)
+			if err != nil {
+				// Try to login again with app-ID and app-Secret
+				if secret, err = v.login(); err != nil { // failed -> try again
+					const retryDelay = 1 * time.Minute
```

Comment:
Move the const up, defining it here is not correct. Atleast it should be right after`authenticate()`

Reply:
Sry if bikeshedding, but what is the rationale here?! There is no point in extending the scope of that value.
Notice that this was a variable (which should never change) before. The current code is equal to `time.Sleep(1*time.Minute)` - but `1*time.Minute` has a explicit name (for readability) (-> no magic values).

Reply:
Yeah but re-definition of a const inside a go-routine is not correct here, if it's a constant then it doesn't change then why define it again and again. If it changes then its not a constant. Either it should be a binomial retry or simply `time.Sleep(1 * time.Minute)` over having `const` definition which doesn't add any value AFAICS.

Reply:
>Yeah but re-definition of a const inside a go-routine is not correct here, if it's a constant then it doesn't change then why define it again and again. If it changes then its not a constant.

Not sure whether I understand you correctly... The constant does not change - it's a (unexported) constant. AFAICS the go binary will look equal - regardless whether you write `const retryDelay = 1 *time.Minute` + `time.Sleep(retryDelay)` or `time.Sleep(1*time.Minute)`.
It's not like the constant is defined every time the function is called - it's only defined (and assigned) once... However, I kept it b/c of readability and similarity to the previous code. I can change if it actually causes confusion...